src/environments/array.js

```diff
@@ -0,0 +1,475 @@
+            if (!elem) {
+                continue;
+            }
+            const shift = row.pos - offset;
+            elem.depth = row.depth;
+            elem.height = row.height;
+            col.push({type: "elem", elem: elem, shift: shift});
+        }
+
+        col = buildCommon.makeVList(col, "individualShift", null, options);
+        col = makeSpan(
+            ["col-align-" + (colDescr.align || "c")],
+            [col]);
+        cols.push(col);
+
+        if (c < nc - 1 || group.value.hskipBeforeAndAfter) {
+            sepwidth = utils.deflt(colDescr.postgap, arraycolsep);
+            if (sepwidth !== 0) {
+                colSep = makeSpan(["arraycolsep"], []);
+                colSep.style.width = sepwidth + "em";
+                cols.push(colSep);
+            }
+        }
+    }
+    body = makeSpan(["mtable"], cols);
+    return makeSpan(["mord"], [body], options);
+};
+
+const mathmlBuilder = function(group, options) {
+    return new mathMLTree.MathNode(
+        "mtable", group.value.body.map(function(row) {
+            return new mathMLTree.MathNode(
+                "mtr", row.map(function(cell) {
+                    return new mathMLTree.MathNode(
+                        "mtd", [mml.buildGroup(cell, options)]);
+                }));
+        }));
+};
+
+// Arrays are part of LaTeX, defined in lttab.dtx so its documentation
+// is part of the source2e.pdf file of LaTeX2e source documentation.
+// {darray} is an {array} environment where cells are set in \displaystyle,
+// as defined in nccmath.sty.
+defineEnvironment({
+    type: "array",
+    names: ["array", "darray"],
+    props: {
+        numArgs: 1,
+    },
+    handler: function(context, args) {
+        let colalign = args[0];
+        colalign = colalign.value.map ? colalign.value : [colalign];
+        const cols = colalign.map(function(node) {
+            const ca = node.value;
+            if ("lcr".indexOf(ca) !== -1) {
+                return {
+                    type: "align",
+                    align: ca,
+                };
+            } else if (ca === "|") {
+                return {
+                    type: "separator",
+                    separator: "|",
+                };
+            }
+            throw new ParseError(
+                "Unknown column alignment: " + node.value,
+                node);
+        });
+        let res = {
+            type: "array",
+            cols: cols,
+            hskipBeforeAndAfter: true, // \@preamble in lttab.dtx
+        };
+        res = parseArray(context.parser, res, dCellStyle(context.envName));
+        return res;
+    },
+    htmlBuilder,
+    mathmlBuilder,
+});
+
+// The matrix environments of amsmath builds on the array environment
+// of LaTeX, which is discussed above.
+defineEnvironment({
+    type: "array",
+    names: [
+        "matrix",
+        "pmatrix",
+        "bmatrix",
+        "Bmatrix",
+        "vmatrix",
+        "Vmatrix",
+    ],
+    props: {},
```

Comment:
Should `props` be optional?

Reply:
That's a good question.  `defineFunction` requires `numArgs` on `props` even if it's `0`.  I'm not sure how useful that is, but I'd rather be consistent so I'll make it mandator here too.